rfc/external-kms.md

```diff
@@ -0,0 +1,113 @@
+
+## Details on status quo
+
+The CLI generates the master secret and passes it to the Bootstrapper on init.
+The CLI also injects it into the Helm charts as a Kubernetes secret for the Constellation KMS service (referred to as CKMS in the following to avoid confusion).
+
+The Bootstrapper uses the master secret to
+
+* derive the *measurement salt* and the *clusterID*.
+* derive the DEK and set up the state disk.
+
+Note that the Bootstrapper performs these operations itself.
+Particularly, it replicates the DEK derivation of the CKMS in ClusterKMS mode.
+
+The CKMS serves a gRPC API to get DEKs.
+It's designed to support different (cloud) KMS services as backend.
+Currently, only the ClusterKMS backend is used, which uses the master secret from the Kubernetes secrets to derive the DEKs.
+The clients of the CKMS are the join-service and the CSI drivers.
+
+### Implemented, but yet unused features
+
+There are CKMS backends for Azure, GCP, and AWS.
+These should be working, but aren't battle-tested.
+
+The init gRPC API has the following fields:
+
+```
+  bytes master_secret = 2;
+  string kms_uri = 3;
+  string storage_uri = 4;
+  string key_encryption_key_id = 5;
+  bool use_existing_kek = 6;
+  bytes salt = 10;
+```
+
+* `master_secret` (and `salt`) are used as described above. The other fields are unused.
+* `kms_uri` and `storage_uri` contain the type and configuration of the (external) KMS and storage to use. The CKMS already has logic to parse them and create corresponding backends.
+* `key_encryption_key_id` and `use_existing_kek` are supposed to control how the external KMS is used.
+
+## Overview of Hashicorp Vault's KMS integrations
+
+Multiple features of Vault integrate with KMSs.
+
+### KMS secrets engine
+
+A KMS secrets engine uses an external KMS to perform operations like encryption and decryption.
+Vault provides a [KMS engine for GCP](https://developer.hashicorp.com/vault/docs/secrets/gcpkms) only.
+
+### Auto unseal
+
+When you start Vault, it's in a sealed state.
+You must provide the seal key to unseal it.
+This is similar to recovering a Constellation cluster.
+
+Instead of manually providing the key, you can configure Vault to [automatically use a KMS to decrypt its root key](https://developer.hashicorp.com/vault/docs/concepts/seal#auto-unseal).
+You can use Azure, GCP, and AWS KMSs among others, but not KMIP.
+
+Vault enterprise features [seal wrap](https://developer.hashicorp.com/vault/docs/enterprise/sealwrap).
+This is an "extra layer of protection" that uses a KMS to encrypt single values before storing them in Vault.
+The docs promote this as a means of meeting compliance requirements rather than improving security.
+
+### KMIP server
+
+Vault can act as a [KMIP server](https://developer.hashicorp.com/vault/docs/secrets/kmip).
+This means that clients can connect to Vault via both the native API and a KMIP API.
+Vault does *not* provide any KMIP client functionality.
```

Comment:
Maybe I understood the KMIP requirement wrong. How would we use a Vault implementation of a KMIP client? 

I was imagining that we use Vault as the KMIP server for our Constellation KMIP client (a service that runs inside Constellation). This way we have a simple software KMIP server to implement against (you can bootstrap Vault locally for testing).
Customers can choose to also use Vault Enterprise as the KMIP server (my hypothesis is this attractive since many companies already run Vault), or anything else that implements KMIP server, e.g., [Thales HSM](https://cpl.thalesgroup.com/encryption/key-management-interoperability-protocol-kmip).

Reply:
Oh, I misunderstood you in this regard. I thought you suggested using Vault as a "proxy" that relays requests to a KMIP server. So we would only need to implement against the Vault API instead of the more complex KMIP API.

I agree that Vault could be a good choice for the first KMIP server to implement against. However, the cloud offering doesn't seem to have that feature and the Enterprise edition requires to "contact sales", which may be annoying.

Reply:
The **development option** seems to have all the features, but none of the SLAs and seems like a great low-cost solution for us developing this feature: https://cloud.hashicorp.com/products/vault/pricing

Reply:
FAQ on that same page says
> Which Vault Enterprise features are not yet available in HCP Vault?
The **Advanced Data Module (ADP)**, resource limiting, and 3rd party plug-ins are not yet available in HCP Vault.

And KMIP is part of ADP.

Reply:
Ah, I had the same misunderstanding.
Using Vault as our KMIP server for dev and testing makes sense to me even if we have to pay the 20$/month license fee.